src/templateAccessibilityLabelForRule.ts

```diff
@@ -0,0 +1,111 @@
+import { BoundDirectivePropertyAst, ElementAst } from '@angular/compiler';
+import { sprintf } from 'sprintf-js';
+import { IRuleMetadata, RuleFailure, Rules, Utils } from 'tslint/lib';
+import { SourceFile } from 'typescript/lib/typescript';
+import { NgWalker } from './angular/ngWalker';
+import { BasicTemplateAstVisitor } from './angular/templates/basicTemplateAstVisitor';
+import { mayContainChildComponent } from './util/mayContainChildComponent';
+
+interface ILabelForOptions {
+  depth: number;
+  labelComponents: string[];
+  labelAttributes: string[];
+  controlComponents: string[];
+}
+export class Rule extends Rules.AbstractRule {
+  static readonly metadata: IRuleMetadata = {
+    description: 'Checks if the label has associated for attribute or a form element',
+    optionExamples: [
+      [true, { depth: 3, labelComponents: ['app-label'], labelAttributes: ['id'], controlComponents: ['app-input', 'app-select'] }]
+    ],
+    options: {
+      items: {
+        type: 'object',
+        properties: {
+          depth: {
+            type: 'number'
+          },
+          labelComponents: {
+            type: 'array',
+            items: {
+              type: 'string'
+            }
+          },
+          labelAttributes: {
+            type: 'array',
+            items: {
+              type: 'string'
+            }
+          },
+          controlComponents: {
+            type: 'array',
+            items: {
+              type: 'string'
+            }
+          }
+        }
+      },
+      type: 'array'
+    },
+    optionsDescription: 'Determine the depth of nesting for form control, custom label, label attribute and controls',
```

Comment:
By default we would support depth of 2, so if someone does 

```
<label>
     <span>
          <div><div><input type="text" /></div></div>
    </span>
</label>
```

it would fail, because the input is at depth 3, but if they add a depth which could be between 1 and 25, then the above code will pass

Reply:
What'a the rationale behind the `depth` option?

Reply:
We want to make sure that if the user is using the label incorrectly in his project, if he uses label and has a huge dom tree inside it, a lot of times in his project, then we do not want to slow this rule from running, and also to keep it fast we have default of 2, and if the user uses a depth for some styling purpose, then we do not want him to stop using this rule, we want him to be able to bump up the depth a bit for his needs

Reply:
I think it's safe to skip the depth check. The changes are that the level of nesting will not have performance implications for the rule.

Reply:
Sometimes developers tend to have huge template files, but anyways, should I get rid of depth altogether or just as a option, if I were to just remove it as a option, what do you think the depth should be?

Reply:
Walking the entire template is fine, no need to set any restrictions.